lambdaclass/mirra_backend
# [GH-800] Add active feature to quests

>[!Warning]
> Merge with [gh-1971-add-quest-feature](https://github.com/lambdaclass/champions_of_mirra/pull/2003)
## Motivation

We need to rework how we communicate and complete daily quests, also add weekly quests:
- We need to send the state and information of the quest that a user has in an endpoint, we'll the user show endpoint for this purpose
- implement weekly quests they are the same as daily but you have a whole week to complete them instead of just one day
- Currently, we're completing the daily quest when a match is finished, this is wrong. The daily quests should be claimed through a button in the client.
- We need to generate quests when either:
  -  We create an user
  - A user logs in and it has passed a day since the quests were generated. We'll check if the day have changed, not counting time

The specs for daily quests now goes like:
- Now a user will begin the day with 6 quests:
  - 3 deactivated and 3 activated
- We'll only count progress for matches that happened on the same day of that quest creation and that happened AFTER the quest was activated
- When you claim a quest you'll activate one of the inactive quests and activate it

Closes #800 
Closes #796

## Summary of changes

The main objective of this pr is to be able to fully use the quest feature without accessing the elixir terminal, and also implement the weekly quests feature

- Removed quest completion on match registration and added endpoint to complete/claim quests
- Added quests to the user when retrieving him through the show endpoint
- Added weekly quest type
- Add `activated_at` for `user_quest`
- Add `last_daily_quest_generation` to users
- Add inactive quest activation on quest completion
- Add 3 new quests to meet the 6 quest minimum

## How to test it?

- Enter the game as a guest in Unity and copy the user_id:
<img width="664" alt="Captura de pantalla 2024-07-24 a la(s) 5 10 15 p  m" src="https://github.com/user-attachments/assets/bef03a34-20ce-4220-8227-5b1f11ac6d03">

- Get to see if he has any quest
`curl -X GET  http://localhost:4001/curse/users/user_id`

- Play a match in unity and lose 

- if you run the get endpoint again you should see that the quest has some progress depending on your actions in the match, there's a quest that you need to lose a match to complete, copy that quest "id" field and run the following command:
`curl -X GET  http://localhost:4001/curse/users/user_id/quest/quest_id/complete_quest`

- Get the user and check that the quest is finished and a new quest turn actives, this means that it has a value in `activated_at`
`curl -X GET  http://localhost:4001/curse/users/user_id`

## Checklist
